File: frontend/src/types/simulation.ts
export interface CompoundInterestRequest {
  initial_amount: string
  monthly_amount: string
  annual_rate_pct: string
  months: number
}

export interface CompoundInterestPoint {
  month: number
  invested: string
  total: string
  interest: string
}

export interface CompoundInterestResponse {
  points: CompoundInterestPoint[]
  total_invested: string
  total_interest: string
  final_amount: string
}

export interface AssetComparisonRequest {
  tickers: string[]
  benchmark_codes: string[]
  start: string
  end?: string | null
}

export interface ComparisonPoint {
  date: string
  value: string
}

export interface ComparisonSeries {
  key: string
  label: string
  points: ComparisonPoint[]
}

export interface AssetComparisonResponse {
  series: ComparisonSeries[]
}

File: backend/app/schemas/simulation.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator


class CompoundInterestRequest(BaseModel):
    initial_amount: Decimal = Field(ge=0)
    monthly_amount: Decimal = Field(ge=0)
    annual_rate_pct: Decimal
    months: int = Field(gt=0, le=600)


class CompoundInterestPoint(BaseModel):
    month: int
    invested: Decimal
    total: Decimal
    interest: Decimal


class CompoundInterestResponse(BaseModel):
    points: list[CompoundInterestPoint]
    total_invested: Decimal
    total_interest: Decimal
    final_amount: Decimal


class AssetComparisonRequest(BaseModel):
    tickers: list[str] = Field(default_factory=list)
    benchmark_codes: list[str] = Field(default_factory=list)
    start: date
    end: date | None = None

    @field_validator("benchmark_codes")
    @classmethod
    def validate_benchmark_codes(cls, value: list[str]) -> list[str]:
        allowed = {"cdi", "ipca", "ibov"}
        for code in value:
            if code not in allowed:
                raise ValueError(f"Indexador inválido: {code}")
        return value


class ComparisonPoint(BaseModel):
    date: date
    value: Decimal


class ComparisonSeries(BaseModel):
    key: str
    label: str
    points: list[ComparisonPoint]


class AssetComparisonResponse(BaseModel):
    series: list[ComparisonSeries]
